AsyncPages: Stop reading after err

joe-elliott opened the pull request:
Putting this PR up for discussion. Tempo currently rarely panics with a stack trace like the following. We have found this quite difficult to reproduce. 

```
	/Users/joe/dev/joe-elliott/tempo/vendor/github.com/parquet-go/parquet-go/page.go:129 +0xfd
created by github.com/parquet-go/parquet-go.AsyncPages in goroutine 610894
	/Users/joe/dev/joe-elliott/tempo/vendor/github.com/parquet-go/parquet-go/page.go:269 +0x178
github.com/parquet-go/parquet-go.readPages({0x3fdf8f0, 0xc0bf919ae0}, 0xc10c1dc310, 0xc095e75880, 0x4007b18?, 0xc10c1dc3f0)
	/Users/joe/dev/joe-elliott/tempo/vendor/github.com/parquet-go/parquet-go/file.go:801 +0x9f
github.com/parquet-go/parquet-go.(*FilePages).ReadPage(0xc0bf919ae0)
	/Users/joe/dev/joe-elliott/tempo/vendor/github.com/parquet-go/parquet-go/file.go:916 +0x65
github.com/parquet-go/parquet-go.(*FilePages).readPage(0xc0bf919ae0, 0xc07d7b5740, 0xc096ba61e0)
	/Users/joe/dev/joe-elliott/tempo/vendor/github.com/parquet-go/parquet-go/buffer.go:534 +0x12d
github.com/parquet-go/parquet-go.(*bufferPool).get(0x623bf20?, 0xfffffffffffffffc?)
goroutine 611543 [running]:

panic: runtime error: slice bounds out of range [:-4]
```

On the surface this is failing b/c the [header decoding](https://github.com/parquet-go/parquet-go/blob/ba969aa4929b45459259ea5c2c4a09876a881c04/file.go#L798) is not returning an error, but is returning an invalid page with a negative value in the compressed page size. The dump of the decoded page header struct is the following. We can see that essentially all values are zero except for a -4 in the CompressedPageSize. I am confident this is not due to a corrupt parquet file. We would be able to reproduce with extreme consistency in that case.

```
&{DATA_PAGE 0 -4 0 <nil> <nil> <nil> <nil>}
```

I believe this is due to AsyncPages calling ReadPage on the wrapped page object after it has already returned an error. We regularly issue jobs and cancel their context if the conditions of a search are satisfied. It is trivial with some debug logging to show calls of ReadPages after it has previously returned an error. 

I have not worked out the actual mechanics by which this occurs but reviewing the code in FilePages it does seem possible for it to be in an unknown state after it returns an error which could potentially result in a panic like the above.

I'm concerned about the impact of this change and would like to discuss. io.EOF handling is always tricky and I don't know if this will have unintended consequences in other codebases.

achille-roussel reviewed (commented):
If I'm understanding, we're calling either SeekToRow or ReadPage on the underlying Page after it returned io.EOF and that's causing the kind of reuse that invalidates the state?

The fix is meaningful to me, tho I wonder if we could handle it more simply by checking for io.EOF after the select?

If FilePage gets corrupted when used after io.EOF, it seems worth fixing too, but totally OK if we focus on one problem at a time. 

I don't have concerns, if our tests pass then that's the extent of our promises regarding backward compatibility 👍 and you've added a test to verify the previously untested behavior, thanks for that!

joe-elliott commented:
> If I'm understanding, we're calling either SeekToRow or ReadPage on the underlying Page after it returned io.EOF and that's causing the kind of reuse that invalidates the state?

The proposed change is to stop calling SeekToRow or ReadPage if either return an error that is not nil and not io.EOF. I excluded io.EOF b/c it's such a strange error in Go and I wasn't sure if we should continue allow reading the wrapped Pages object or not. So this would stop making call to the Pages object after it received what it considers a "fatal" error.

In Tempo we often execute many thousands of simultaneous jobs and then cancel them if the search conditions were satisfied elsewhere. So in our case the error in question was actually `context.Canceled`. Since writing this PR I believe I have pieced together our panic. 

Many Bothans died for this:

1. While [decoding a header](https://github.com/parquet-go/parquet-go/blob/main/file.go#L798) in FilePages.ReadPage the thrift code will read short segments and sometimes [individual bytes](https://github.com/parquet-go/parquet-go/blob/main/encoding/thrift/binary.go#L207).
1. One of these very short reads misses the [bufio.Reader's](https://github.com/parquet-go/parquet-go/blob/ba969aa4929b45459259ea5c2c4a09876a881c04/file.go#L759) internal buffer triggering a read from the backend.
1. [Cache misses](https://github.com/grafana/tempo/blob/4ab3c173af81c037af9e5d77f30309e78d57eed7/tempodb/backend/cache/cache.go#L116) and the backend read fails with `context.Canceled` b/c the job was canceled for any number of reasons.
1. Something else successfully reads that exact parquet page and stores it in cache.
1. The [AsyncPages](https://github.com/parquet-go/parquet-go/blob/ba969aa4929b45459259ea5c2c4a09876a881c04/page.go#L263-L289) loop comes back around and calls `ReadPage` again.
1. [Decoder.decode](https://github.com/parquet-go/parquet-go/blob/main/file.go#L798) is called again but now the underlying buffer is misaligned. We are N bytes into the header instead of at the start of the header. This wouldn't matter if the backend would just return `context.Canceled` again but...
1. This time cache hits and returns successfully because memcached [doesn't honor context](https://github.com/grafana/tempo/blob/4ab3c173af81c037af9e5d77f30309e78d57eed7/pkg/cache/memcached.go#L98-L100).

So this can be solved in Tempo by having the memcached client return context canceled correctly. Up to you if we think this is still worth pursuing in parquet-go.

joe-elliott commented:
I think this can also happen if object storage returns a transient error and then succeeds on the next call.

fpetkovski commented:
Keeping track of the last error and returning it on subsequent calls makes sense to me. FilePages are a transient object - we return a new instance each time we call `Pages()` so I think it's fine to treat them as an iterator which terminates when it encounters an error.

The alternative would be to retry the read which failed, but this assumes the underlying reader can seek back, which I don't think is the case right now.

joe-elliott commented:
@fpetkovski I'm fine with either approach. If we store the last error in `FilePages` and always return it then we may need to review other implementations of `Pages` to see if they need a similar treatment.

@achille-roussel do you have a preferred direction?

achille-roussel commented:
🤔 storing the last fatal error is a good idea generally, tho there are cases that should be valid like:
- read pages until `io.EOF`
- seek to position
- read pages from position

I think `io.EOF` is the only non-fatal error that would allow seeking back to different positions.

joe-elliott commented:
@fpetkovski @achille-roussel if you have a moment can you re-review? i created a Pages implementation that memoizes in any non fatal errors and repeatedly returns them instead of calling the nested pages. This seems in line with what we were discussing above.

wdyt?

achille-roussel commented on page.go:
```diff
@@ -324,6 +326,69 @@ func onePage(page Page) Pages {
+	err := p.pages.SeekToRow(rowIndex)
+	if isFatal(err) {
+		// wrapping the error to add context. if we return a ReadPage() error when SeekToRow is called that could be confusing. this aims to clarify the source of the error
+		p.err = fmt.Errorf("stored SeekToRow err: %w", err)
```
```suggestion
		p.err = fmt.Errorf("unrecoverage error returned by SeekToRow: %w", err)
```

achille-roussel replied:
Yes... I was typing on mobile.

achille-roussel commented on page.go:
```diff
@@ -324,6 +326,69 @@ func onePage(page Page) Pages {
+		return false
+	}
+
+	if err == io.EOF || err == ErrSeekOutOfRange { // FilePages can return ErrSeekOutOfRange but this is recoverable
```
```suggestion
	if err == io.EOF || errors.Is(err, ErrSeekOutOfRange) { // FilePages can return ErrSeekOutOfRange but this is recoverable
```

achille-roussel reviewed (approved):
The approach sound good to me 👍 

I have a personal preference for merging `memoizeErrorPages` into `asyncPages` since we're only using it there and the wrapper is always added... but that's your call, either model works fine.

joe-elliott commented:
Allright! I have reverted to the previous loop where we just store the most recent error and return it from the async pages loop. I added ErrSeekOutOfRange as a recoverable error simply b/c it is the way it's implemented in FilePages, but we can remove this if necessary.

wdyt?

achille-roussel reviewed (approved):
:shipit:
